### PR TITLE
[trainer] feat: Enable data shuffling across epochs when ppo_epoch > 1, following MegatronWorker’s approach.

### PR DESCRIPTION
**Pull request by none0663**
### What does this PR do?

Currently, when ppo_epoch > 1 (i.e., multiple optimization steps per batch), the training data is not shuffled between epochs. 
Expected Behavior
As implemented in OpenAI’s [baselines/ppo](https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/ppo2.py#L160), data should be reshuffled at the start of each epoch when ppo_epoch > 1. This ensures diversity in mini-batches and improves model robustness.

Related Issues: https://github.com/volcengine/verl/issues/3213

Proposed Fix

FSDP Implementation: As modified in this PR.
Megatron-LM Implementation: Requires only setting shuffle: True in the YAML configuration and specifying a data_loader_seed.(not need the modified the code in this pr)

### CODE REVIEW

**Comment by vermouth1992** [comment]
Could you follow what MegatronWorker does? https://github.com/volcengine/verl/blob/main/verl/workers/actor/megatron_actor.py#L254

**Comment by none0663** [comment]
> Could you follow what MegatronWorker does? https://github.com/volcengine/verl/blob/main/verl/workers/actor/megatron_actor.py#L254
It seems good; I’ll make the changes soon.

**Comment by techkang** [comment]
@none0663 Any updates on this?

**Comment by none0663** [comment]
> @techkang


I’ve updated the code in the PR and have successfully run and verified the changes on my local machine. Could you please review the updates and run the unit tests when you have a moment?

Let me know if you have any questions or need further adjustments.

Thanks!

**Comment by none0663** [comment]
> Could you follow what MegatronWorker does? https://github.com/volcengine/verl/blob/main/verl/workers/actor/megatron_actor.py#L254

Done

**Comment by techkang** [comment]
LGTM

**Comment by none0663** [comment]
> LGTM

Thinks, I has fixed the bug for the failing checks for` critic, sppo and unit test.`

**Comment by none0663** [comment]
> LGTM
@techkang Could you please review the new updates and run the unit tests  again when you have a moment ?

**Comment by none0663** [comment]
fix the critic  bug and fix  yaml documentation format.

**Comment by techkang** [comment]
Sorry for late, running workflows now.

**Review by vermouth1992 — DISMISSED** [review]

**Comment by none0663** [comment]
Sorry for missing fixing some minor bugs in the critic earlier. They’re now fixed – please re-run the workflows, thanks.

**Comment by none0663** [comment]
@techkang   
The failed checks seem unrelated to the modified code, so I feel like it can be merged into the main branch.

**Comment by techkang** [comment]
> @techkang The failed checks seem unrelated to the modified code, so I feel like it can be merged into the main branch.

I think so. cc @vermouth1992

**Comment by none0663** [comment]
Clipboard_Screenshot_1757782142

**Comment by none0663** [comment]
@vermouth1992 Do I need to resolve this conflict before it can be merged? It appears that the conflict was introduced today.
Clipboard_Screenshot_1757782142

**Comment by none0663** [comment]
> > @techkang The failed checks seem unrelated to the modified code, so I feel like it can be merged into the main branch.
> 
> I think so. cc @vermouth1992
@vermouth1992 
Will this pull request be merged into the main branch? It seems all checks have passed except one that was cancelled.

**Comment by none0663** [comment]
> > @techkang The failed checks seem unrelated to the modified code, so I feel like it can be merged into the main branch.
> 
> I think so. cc @vermouth1992

@vermouth1992 
Hello team,

I noticed that all required checks for [PR #3247] have passed except one that was cancelled. Could you confirm whether this pull request will be merged into the main branch? If so, I’ll proactively resolve any existing conflicts to ensure a smooth merge in the `verl/workers/actor/dp_actor.py`
